src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,645 @@
+                                        var updatedParentExpressionStatement = parentStatement.ReplaceNode(
+                                            source.Parent.Parent.Parent,
+                                            initializer);
+
+                                        documentUpdate = new DocumentUpdate(parentStatement, new[] { updatedParentExpressionStatement, qe1 });
+                                        return true;
+                                    }
+                                }
+                            }
+
+                            break;
+                        case SyntaxKind.ReturnStatement:
+                            string freeVariableName = GetFreeSymbolName(variableName, source.GetLocation().SourceSpan.Start);
+                            var identifierName = SyntaxFactory.IdentifierName(freeVariableName);
+                            if (TryConvertQueryExpression(source, (ExpressionSyntax expression) => leafExpressionCreationMethod(identifierName, expression), out var qe2))
+                            {
+                                if (TryFindParentAssignment(source, out var parentStatement))
+                                {
+                                    var variableDeclaration =
+                                        SyntaxFactory.LocalDeclarationStatement(
+                                            SyntaxFactory.VariableDeclaration(
+                                                VarNameIdentifier,
+                                                SyntaxFactory.SingletonSeparatedList(
+                                                    SyntaxFactory.VariableDeclarator(
+                                                        identifierName.Identifier,
+                                                        null,
+                                                        SyntaxFactory.EqualsValueClause(initializer)))));
+                                    var returnList = SyntaxFactory.ReturnStatement(identifierName);
+                                    documentUpdate = new DocumentUpdate(parentStatement, new[] { variableDeclaration, qe2, returnList });
+                                    return true;
+                                }
+                            }
+
+                            break;
+                        case SyntaxKind.ArrowExpressionClause:
+                            // Not supported
+                            break;
+                        default:
+                            throw new NotImplementedException(parenthesizedExpression.Parent.Parent.Parent.Kind().ToString());
+                    }
+                    documentUpdate = default;
+                    return false;
+                }
+
+                throw new NotImplementedException(parenthesizedExpression.Parent.Parent.Kind().ToString());
+            }
+
+            private bool TryGetVariableName(SyntaxNode node, out ExpressionSyntax identifier)
```

Comment:
calling this 'identifier' is very strange given it may not be an identifier right?

Reply:
agree. changed to 'variable'. please let me know if you see a better name.

---
In reply to: [175267744](https://github.com/dotnet/roslyn/pull/25362#discussion_r175267744) [](ancestors = 175267744)